### DISCUSSION TITLE
Unstable bug about project0 using java

### DISCUSSION THREAD

**Discussion by Floooorian-shi** [sist-cs120/project-wiki#36] [General]
In 04_java_sound_player, I am not sure what does "(1) Move or copy the library to C:\WINDOWS\system32. This is the default search location for JNI libraries. (2) Inform the JVM where the library is located. This can be done with, e.g. java -Djava.library.path=C:\WINDOWS\system32." mean. I tried to complete its steps. 
However, in AudioHw.java, the line "import com.synthbot.jasiohost.*;" said "程序包com.synthbot.jasiohost不存在". 

When I compile it, it said "Exception in thread "main" com.synthbot.jasiohost.AsioException: No input or output is present.", so that the code "audiohw.init()" failed. But in afternoon, I didn't meet this problem, and the example code can be compiled well, although I can't hear the sinewave sound.
![QQ20240928-232014](https://github.com/user-attachments/assets/45acd13d-f206-4d86-b0aa-89055b58480b)

**Comment by SIST1C407** [comment]
1. The two operations mentioned in the wiki both allow java to find `jasiohost.dll` in `java.library.path` to support the operation of `loadLibrary` in `JAsioHost.jar`. The fundamental way to achieve this purpose is to change `java.library.path`. There are two ways to change `java.library.path`:
   1. You can choose to move `jasiohost.dll` to some paths that `java.library.path` searches for by default as the first way mentioned in the wiki. For example, in windows it is the paths in the `PATH` environment variable, and in linux it is the paths in the `LD_LIBRARY_PATH` environment variable. A common practice for this approach is adding the location of the `jasiohost.dll` to the environment variable instead of copying the it directly to the `C:\WINDOWS\system32` folder, for which we'll modify the wiki a bit.
   2. You can also choose to specify the `java.library.path` when compile the source file with the java program directly on the command line as the second way mentioned in the wiki.
2. About the error *程序包com.synthbot.jasiohost不存在*, you should add the file `doc/04_java_sound_player/lib/JAsioHost.jar` in repo to `<your java project source path>/lib`. So the jar can be found by the java compiler.
![image](https://github.com/user-attachments/assets/aac72a46-457b-49ca-bf02-27164aa84ebe)

3. Your problem is not an error due to an `UnsatisfiedLinkError` triggered by the location of `jasiohost.dll`. I think you should prepare the [Step 1](https://github.com/CALMorACT/project-wiki/tree/main/doc/01_win_asio_juce) carefully to ensure the ASIO4ALL driver right. If you have any questions, please feel free to ask me. Or, you can try change the index used for `asioDriver.getChannelOutput` in your `AudioHw.init(AudioHw.java)` to see if it works.
